### PR TITLE
feat(headers): add optional header masking feature to outbound request

### PR DESCRIPTION
**Pull request by NishantJoshi00**
## Type of Change

- [ ] Bugfix
- [ ] New feature
- [x] Enhancement
- [ ] Refactoring
- [ ] Dependency updates
- [ ] Documentation
- [ ] CI/CD

## Description

This is an enhanced implementation of optional masking of headers for the outgoing requests (focused mainly on connectors).
This introduces a new type, the allows optional masking on the data that it holds. This type in itself is a enum, that has 2 possible values, `Masked` and `Normal`. `new_masked` method is used, then a masked value is created. Use of `Into::into` will always create a `Normal` value. This allows ease, of use while still allowing use to provide the feature to mask certain fields.

### Additional Changes

- [ ] This PR modifies the API contract
- [ ] This PR modifies the database schema
- [ ] This PR modifies application configuration/environment variables

## Motivation and Context

Hiding sensitive data from logging framework.

## How did you test it?

Compiler guided & 
Following are the before and after results of this sanitization:
> Before:
Screenshot 2023-05-31 at 2 39 44 AM

> After:
Screenshot 2023-05-31 at 2 32 20 AM

> (PS. The API key that is visible in the before test is not a real key, but generated by an RNG)

## Checklist

- [x] I formatted the code `cargo +nightly fmt --all`
- [x] I addressed lints thrown by `cargo clippy`
- [x] I reviewed submitted code
- [ ] I added unit tests for my changes where possible
- [ ] I added a [CHANGELOG](/CHANGELOG.md) entry if applicable

### CODE REVIEW

File: crates/router/src/connector/aci.rs [hunk 1]
```diff
@@ -35,11 +36,15 @@ impl ConnectorCommon for Aci {
     fn get_auth_header(
         &self,
         auth_type: &types::ConnectorAuthType,
-    ) -> CustomResult<Vec<(String, String)>, errors::ConnectorError> {
+    ) -> CustomResult<Vec<(String, request::OptionalMaskedValue<String>)>, errors::ConnectorError>
```
Comment (dracarys18) [inline crates/router/src/connector/aci.rs:39]: Convert this to a new type maybe use `Headers` type that's already there?

File: crates/router/src/connector/aci.rs [hunk 2]
```diff
@@ -35,11 +36,15 @@ impl ConnectorCommon for Aci {
...
+    {
         let auth: aci::AciAuthType = auth_type
             .try_into()
             .change_context(errors::ConnectorError::FailedToObtainAuthType)?;
-        Ok(vec![(headers::AUTHORIZATION.to_string(), auth.api_key)])
+        Ok(vec![(
+            headers::AUTHORIZATION.to_string(),
+            request::OptionalMaskedValue::new_masked(auth.api_key.into()),
+        )])
```
Comment (dracarys18) [inline crates/router/src/connector/aci.rs:47]: Can you create a  trait that will do this with `into_masked`

File: crates/router/src/services/api/request.rs [hunk 3]
```diff
@@ -1,16 +1,69 @@
...
 
-pub(crate) type Headers = collections::HashSet<(String, String)>;
+pub(crate) type Headers = collections::HashSet<(String, OptionalMaskedValue<String>)>;
+
+///
+/// An Enum that allows us to optionally mask data, based on which enum variant that data is stored
+/// in.
+///
+#[derive(Clone, Eq, PartialEq)]
+pub enum OptionalMaskedValue<T: Eq + PartialEq + Clone> {
```
Comment (dracarys18) [inline crates/router/src/services/api/request.rs:17]: Probably change into `Maskable`

**Review by dracarys18 — CHANGES_REQUESTED** [review]

**Review by dracarys18 — APPROVED** [review]
LGTM

**Review by SanchithHegde — COMMENTED** [review]
Other than that, looks good to me!

File: crates/router/src/connector/nmi.rs
[the remark this answers is not in the document]
Comment (NishantJoshi00) [reply crates/router/src/connector/nmi.rs:1]: Yes please, these changes might vary across various structs and transformers throughout the connectors.

[the remark this answers is not in the document]
Comment (NishantJoshi00) [reply crates/router/src/connector/nmi.rs:1]: https://github.com/juspay/hyperswitch/issues/1348

File: crates/router/src/connector/authorizedotnet.rs [hunk 4]
```diff
@@ -185,7 +186,7 @@ impl ConnectorIntegration<api::PSync, types::PaymentsSyncData, types::PaymentsRe
         &self,
         req: &types::PaymentsSyncRouterData,
         connectors: &settings::Connectors,
-    ) -> CustomResult<Vec<(String, String)>, errors::ConnectorError> {
+    ) -> CustomResult<Vec<(String, request::Maskable<String>)>, errors::ConnectorError> {
         // This connector does not require an auth header, the authentication details are sent in the request body
```
[the remark this answers is not in the document]
Comment (NishantJoshi00) [reply crates/router/src/connector/authorizedotnet.rs:190]: #1348

File: crates/router/src/connector/airwallex.rs [hunk 5]
```diff
@@ -128,11 +132,14 @@ impl ConnectorIntegration<api::AccessTokenAuth, types::AccessTokenRequestData, t
...
         let headers = vec![
-            (headers::X_API_KEY.to_string(), req.request.app_id.clone()),
-            ("Content-Length".to_string(), "0".to_string()),
-            ("x-client-id".to_string(), req.get_request_id()?),
+            (
+                headers::X_API_KEY.to_string(),
+                req.request.app_id.clone().into(),
+            ),
+            ("Content-Length".to_string(), "0".to_string().into()),
+            ("x-client-id".to_string(), req.get_request_id()?.into()),
```
[the remark this answers is not in the document]
Comment (NishantJoshi00) [reply crates/router/src/connector/airwallex.rs:142]: I agree, doing the change now!

**Review by SanchithHegde — APPROVED** [review]

**Review by ashokkjag — APPROVED** [review]

**Pull request merged**